server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java

```diff
@@ -0,0 +1,146 @@
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.api.server.ws.WebService.NewAction;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.component.index.ComponentIndex;
+import org.sonar.server.component.index.ComponentIndexQuery;
+import org.sonarqube.ws.WsComponents.Component;
+import org.sonarqube.ws.WsComponents.SuggestionsWsResponse;
+import org.sonarqube.ws.WsComponents.SuggestionsWsResponse.Qualifier;
+
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.component.ComponentsWsParameters.ACTION_SUGGESTIONS;
+
+public class SuggestionsAction implements ComponentsWsAction {
+
+  private static final String URL_PARAM_QUERY = "s";
+
+  private static final String[] QUALIFIERS = {
+    Qualifiers.VIEW,
+    Qualifiers.SUBVIEW,
+    Qualifiers.PROJECT,
+    Qualifiers.MODULE,
+    Qualifiers.FILE,
+    Qualifiers.UNIT_TEST_FILE
+  };
+
+  private static final int NUMBER_OF_RESULTS_PER_QUALIFIER = 6;
+
+  private final ComponentIndex index;
+
+  private DbClient dbClient;
+
+  public SuggestionsAction(DbClient dbClient, ComponentIndex index) {
+    this.dbClient = dbClient;
+    this.index = index;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    NewAction action = context.createAction(ACTION_SUGGESTIONS)
+      .setDescription("Internal WS for the top-right search engine")
+      .setSince("4.2")
+      .setInternal(true)
+      .setHandler(this)
+      .setResponseExample(Resources.getResource(this.getClass(), "components-example-suggestions.json"));
+
+    action.createParam(URL_PARAM_QUERY)
+      .setRequired(true)
+      .setDescription("Substring of project key (minimum 2 characters)")
+      .setExampleValue("sonar");
+  }
+
+  @Override
+  public void handle(Request wsRequest, Response wsResponse) throws Exception {
+    SuggestionsWsResponse searchWsResponse = doHandle(wsRequest.param(URL_PARAM_QUERY));
+    writeProtobuf(searchWsResponse, wsRequest, wsResponse);
+  }
+
+  SuggestionsWsResponse doHandle(String query) {
+    List<Qualifier> resultsPerQualifier = getResultsOfAllQualifiers(query);
+
+    return SuggestionsWsResponse.newBuilder()
+      .addAllResults(resultsPerQualifier)
+      .build();
+  }
+
+  private List<Qualifier> getResultsOfAllQualifiers(String query) {
+    return Arrays
+      .stream(QUALIFIERS)
+      .flatMap(qualifier -> getResultsOfQualifier(query, qualifier).map(Stream::of).orElseGet(Stream::empty))
+      .collect(Collectors.toList());
+  }
+
+  private Optional<Qualifier> getResultsOfQualifier(String query, String qualifier) {
+    ComponentIndexQuery componentIndexQuery = new ComponentIndexQuery(query)
+      .setQualifier(qualifier)
+      .setLimit(NUMBER_OF_RESULTS_PER_QUALIFIER);
+
+    List<String> uuids = searchInIndex(componentIndexQuery);
+    if (uuids.isEmpty())
```

Comment:
We enforce usage of `{}` around all blocks, even if it contains a single statement.